event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -41,30 +51,123 @@
 
 	@EventListener
-	public void onFunctionRegistered(ResourceAddedEvent<XFunction> event) {
+	public void onFunctionRegistered(ResourceAddedOrModifiedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
 		this.functions.put(functionName, functionResource);
-		this.launcher.dispatch(functionResource);
+		this.kafkaConsumerMonitor.beginTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		if (!running) {
+			running = true;
+			scalingThread.start();
+		}
+		synchronized (mutex) {
+			this.mutex.notify();
+		}
+
 		logger.info("function added: " + functionName);
 	}
 
+
 	@EventListener
 	public void onFunctionUnregistered(ResourceDeletedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
-		this.kubernetesClient.extensions().jobs().withLabel("function", functionName).delete();
 		this.functions.remove(functionName);
+		deployer.undeploy(functionResource);
+
+		this.kafkaConsumerMonitor.stopTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		synchronized (mutex) {
+			this.mutex.notify();
+		}
+
 		logger.info("function deleted: " + functionName);
 	}
+
+	@PreDestroy
+	public void tearDown() {
+		running = false;
+		synchronized (mutex) {
+			mutex.notify();
+		}
+		kafkaConsumerMonitor.close();
+	}
+
+	/**
+	 * Updates the wake up interval of the scaling thread to be the minimal desired value of all
+	 * functions, or zero (sleep forever) if there is no function.
+	 */
+	private void updateWakeUpInterval() {
+		this.scalingFrequencyMs = this.functions.values().stream()
+				.mapToInt(f -> f.getSpec().getScalingStrategy() == null ? 1_000
+						: f.getSpec().getScalingStrategy().getMaxUpdatePeriodMilliSeconds())
+				.min().orElse(0);
+	}
+
+	private class ScalingMonitor implements Runnable {
+
+		@Override
+		public void run() {
+			while (running) {
+
+				Map<KafkaConsumerMonitor.TopicGroup, List<KafkaConsumerMonitor.Offsets>> offsets = kafkaConsumerMonitor.compute();
+				logOffsets(offsets);
+				Map<String, Long> lags = offsets.entrySet().stream()
+						.collect(Collectors.toMap(
+							e -> e.getKey().group,
+							e -> e.getValue().stream()
+									.mapToLong(KafkaConsumerMonitor.Offsets::getLag)
+									.max().getAsLong()
+						));
+				functions.values().parallelStream().forEach(
+						f -> {
+							String expression = f.getSpec().getScalingStrategy() != null ? f.getSpec().getScalingStrategy().getReplicasExpression() : "1";
+							Expression spel = spelExpressionParser.parseExpression(expression);
+							EvaluationContext context = new StandardEvaluationContext();
+							context.setVariable("lag", lags.get(f.getMetadata().getName()));
+							int replicas = spel.getValue(context, Integer.class);
+							// TODO no more replicas than topic.partitions
```

Comment:
I feel like we need this ceiling enforcement before merging. Otherwise, it's too easy to spin up excess instances.